js/ui/status/keyboard.js

```diff
@@ -480,6 +481,48 @@ const InputSourceManager = new Lang.Class({
             this._updateMruSettings();
     },
 
+    passwordMode: function(enable) {
+        let source;
+
+        if (enable) {
+            // do not switch the keyboard layout if the current one is a latin one
+            if (this._keyboardManager.isLatinLayout(this._currentSource.id))
+                return;
+
+            // look for latin keyboard layout to switch to
+            for (let i = 0; i < this._mruSources.length; ++i) {
+                let src = this._mruSources[i];
+                if (this._keyboardManager.isLatinLayout(src.id)) {
+                    source = src;
+                    break;
+                }
+            }
+            if (!source) {
+                // fallback to an us keyboard layout
+                source = new InputSource('xkb',
+                                         'us',
+                                         'English (US)',
+                                         'en',
+                                         Object.keys(this._inputSources).length);
```

Comment:
Instead of doing all this work here (updating mru sources, emitting signals...) I'd save this new source as e.g. `this._fallbackInputSource`, then call `this._inputSourcesChanged()` and add code in there that adds it to `this._inputSources` when set.
In other words, the only place that changes `this._inputSources` should be `this._inputSourcesChanged()`.

Reply:
Much nicer indeed! Thanks for the hint, adjusted accordingly.

Reply:
I would go one step further, and move the creation of this source also in `this._inputSourceChanged()`; you could simply add/clear a flag in this function that is e.g. `this._needsFallbackSource`. That way you would not need to look at the length of `this._inputSources` keys from here.